### DISCUSSION TITLE
Incomplete type for `log_config` parameter to `uvicorn.run()` in 0.18.1

### DISCUSSION THREAD

**Discussion by wch** [Kludex/uvicorn#1536] [Potential Issue] [answered]
The type for the `log_config` parameter in `run()` is this:

```py
    log_config: typing.Optional[typing.Union[dict, str]] = None,
```

https://github.com/encode/uvicorn/blob/11d28e208504436e7e06603a941c01b76f54346f/uvicorn/main.py#L481

I believe it should be 

```py
    log_config: typing.Optional[typing.Union[typing.Dict[str, typing.Any], str]] = None,
```

mypy seems to be OK with the existing code, but pyright flags the current code as an error, with the following message:

```
error: Type of "run" is partially unknown
``` 

For a simple test of how mypy and pyright handle a bare `dict` type:

```
$ mypy -c "x: dict = {}"  
Success: no issues found in 1 source file
```

```
$ echo "x: dict = {}" > test.py
$ pyright test.py
.... <output removed>
Searching for source files
Found 1 source file
pyright 1.1.255
.../test.py
  .../test.py:1:4 - error: Expected type arguments for generic class "dict" (reportMissingTypeArgument)
 .../test.py:1:1 - error: Type of "x" is partially unknown
    Type of "x" is "dict[Unknown, Unknown]" (reportUnknownVariableType)
2 errors, 0 warnings, 0 informations 
Completed in 0.511sec
```

It's possible that mypy is just less strict, but in any case, a stricter type would be helpful.

The same thing goes for the `log_config` parameter in the `Config` class.

**Comment by Kludex** [comment, marked as answer]
Sure. PR welcome to use a more explicit type annotation.

Not sure if it was a suggestion, but I think that changing the type checker is out of question. 😅

**Comment by wch** [comment]
Ha, no, I wasn't suggesting changing the type checker. It's just that I use pyright on my code which calls `uvicorn.run()`, and it's flagging this issue when I do that.

**Reply by Kludex** [reply]
No problem. But yeah, PR is welcome to improve the type annotation on the ones you mentioned. 🙏
